### ISSUE TITLE
Finetuning example?

### ISSUE THREAD

**Issue by vmajor** [state-spaces/mamba#15]
I see good amount of focus on understanding how to perform full training of Mamba, but what about PEFT? Adapters/LoRA finetuning.

The base models are in fact "Ready" for a fine tune, however due to the architecture I would prefer to minimize trial and error and ask: what can I use to fine tune a pretrained Mamba model?

Do you have a minimal code example?

**Comment by albertfgu** [comment]
We have not tried finetuning any models yet. We're also very interested in understanding their performance downstream! In principle it should be possible to finetune Mamba the same way that you would any other LLM.

**Comment by geronimi73** [comment]
> Do you have a minimal code example?

here's a minimal example that works for me: https://github.com/geronimi73/mamba/blob/main/finetune.py
mixer_seq_simple.py has to be changed to take labels as input: https://github.com/geronimi73/mamba/blob/1f728674169523ba76b7158503c1565a4819ca3b/mamba_ssm/models/mixer_seq_simple.py#L226

the finetuned model "works" but still probably lots of things wrong, it's a first try

**Comment by erenirmak** [comment]
> We have not tried finetuning any models yet. We're also very interested in understanding their performance downstream! In principle it should be possible to finetune Mamba the same way that you would any other LLM.

will you add other downstream task heads ?

**Comment by justusmattern27** [comment]
Hey everyone - we've managed to train mamba by modifying the Huggingface Trainer class. Here is our [implementation](https://github.com/havenhq/mamba-chat/tree/main), we were actually able to train a chat model that seems to perform quite well. 

Of course, performance is still miles away from Mistral and Llama, but quite impressive given the small model size and how new the apporach is. Awesome work! @albertfgu @tridao
